src/pattern_matching/pssm/dnamotif.rs

```diff
@@ -0,0 +1,269 @@
+// Copyright 2018 Kieran Hervold
+// Licensed under the MIT license (http://opensource.org/licenses/MIT)
+// This file may not be copied, modified, or distributed
+// except according to those terms.
+
+use super::*;
+use ndarray::prelude::Array2;
+use std::f32;
+use std::f32::{INFINITY, NEG_INFINITY};
+
+#[derive(Clone, Debug, PartialEq)]
+pub struct DNAMotif {
+    pub seq_ct: usize,
+    pub scores: Array2<f32>,
+    /// sum of "worst" base at each position
+    pub min_score: f32,
+    /// sum of "best" base at each position
+    pub max_score: f32,
+}
+
+impl DNAMotif {
+    pub fn from_seqs_with_pseudocts(
+        seqs: Vec<Vec<u8>>,
+        pseudos: &[f32; 4],
+    ) -> Result<DNAMotif, PSSMError> {
+        // null case
+        if seqs.len() == 0 {
+            return Ok(DNAMotif {
+                seq_ct: 0,
+                scores: Array2::zeros((0, 0)),
+                min_score: 0.0,
+                max_score: 0.0,
+            });
+        }
+        let seqlen = seqs[0].len();
+        let mut counts = Array2::zeros((seqlen, 4));
+        for i in 0..seqlen {
+            for base in 0..4 {
+                counts[[i, base]] = pseudos[base];
+            }
+        }
+
+        for seq in seqs.iter() {
+            if seq.len() != seqlen {
+                return Err(PSSMError::InconsistentLen);
+            }
+
+            for (idx, base) in seq.iter().enumerate() {
+                match Self::lookup(*base) {
+                    Err(e) => return Err(e),
+                    Ok(pos) => counts[[idx, pos]] += 1.0,
+                }
+            }
+        }
+        let mut m = DNAMotif {
+            seq_ct: seqs.len(),
+            scores: counts,
+            min_score: 0.0,
+            max_score: 0.0,
+        };
+        m.normalize();
+        m.calc_minmax();
+        Ok(m)
+    }
+
+    // helper function -- normalize self.scores
+    fn normalize(&mut self) {
+        for i in 0..self.len() {
+            let mut tot: f32 = 0.0;
+            // FIXME: slices would be cleaner
+            for base_i in 0..4 {
+                tot += self.scores[[i, base_i]];
+            }
+            for base_i in 0..4 {
+                self.scores[[i, base_i]] /= tot;
+            }
+        }
+    }
+
+    // helper function
+    pub fn calc_minmax(&mut self) {
```

Comment:
if this is a helper function, I think it should be private. Alternatively, please add a doc comment.